### PR TITLE
Modify Makefile to distinguish between developer and packaging use cases

### PR DESCRIPTION
**Pull request by brson**
## What have you changed? (mandatory)

This PR modifies the makefiles to support two different "release" use cases: optimized builds for development and benchmarking; and builds for actual release and distribution. The two use cases have different requirements. For my purposes I am going to make the "release" targets compile faster while still producing fast code. But there are other distribution-specific requirements that are irrelevant to local development, like verifying SSE4.2 codes and producing the `bin` directory.

This adds new makefile rules for use by CI/CD for making release builds, `dist_release`, `dist_unportable_release`, etc.

The existing "release" rules will be repurposed for developers only.

After this PR lands I will update the SRE repo to only use `dist_*` rules. Then I will change the "release" rules, changing LTO to thinLTO, and removing debuginfo by default.

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Manual.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

Part of https://github.com/tikv/tikv/issues/4674

### CODE REVIEW

**Comment by brson** [comment]
This is the first step to removing the mass of `x-whatever` rules at the bottom of the makefile.

File: Makefile [hunk 1]
```diff
@@ -76,19 +76,41 @@ ctl:
...
 	ROCKSDB_SYS_PORTABLE=0 make release
 
-prof_release:
+# Distributable bins with jemalloc memory profiling
+dist_prof_release:
 	ENABLE_FEATURES=mem-profiling make release
 
-fail_release:
+# Distributable bins instrumented with failpoints
+dist_fail_release:
```
Comment (breeswish) [inline Makefile]: I have a question. When will failpoint release being useful? @BusyJay

Comment (siddontang) [reply Makefile]: For Schrodinger

**Comment by brson** [comment]
ptal @zhouqiang-cl

**Comment by brson** [comment]
I pushed a commit to comment the purpose of fail_release.

**Comment by brson** [comment]
The SRE followup is https://github.com/pingcap/SRE/pull/243

**Comment by brson** [comment]
It doesn't need to happen for this PR, but it has been suggested that the optimized profile must give profilers the info they need and some of that is debuginfo.

I have tested that debuginfo=0 provides symbol names, and I hope that is enough for typical profilers to work, but I am going to test perf with debuginfo=0, opt-level=3 and see if it reports something sensible.

@BusyJay already convinced me that the test profile should be debuginfo=1 to get line numbers in backtraces. It may be the case that profiles would also want line numbers, or that devs doing profiling would want line numbers.

I don't know. In the worst case, I'll merge debuginfo=0, someone will complain, then we'll change it to debuginfo=1.

cc @nrc

**Comment by brson** [comment]
Related to https://github.com/tikv/tikv/pull/4689, that modifies the non-release profiles.

**Comment by brson** [comment]
OK, I think we have had sufficient +1s to go ahead and get this merged. Merging it will not immediately effect SRE/CI/CD since the release targets are temporary dupes of the dist_release targets. So we should be able to just merge this and not worry about the CI.

I've also submitted a cherry-pick of this to release-3.0 https://github.com/tikv/tikv/pull/4841. I am just guessing that is a suitable way to deal with 3.0 changes.

This has been in process for a long time so I'd like to get it moving.

Since @zhouqiang-cl has the full picture of how this affects SRE and how the 3.0 branch works, I think he should give an approval before merging. But PTAL @breeswish @BusyJay @AndreMouche @nrc 

And then do the same on the 3.0 patch: https://github.com/tikv/tikv/pull/4841

File: Makefile [hunk 2]
```diff
@@ -76,19 +76,43 @@ ctl:
 run:
 	cargo run --no-default-features --features  "${ENABLE_FEATURES}" --bin tikv-server
 
+# An optimized build suitable for development and benchmarking
 release:
+	make dist_release
+
+# An optimized build with additional SSE4.2 optimizations
```
Comment (nrc) [inline Makefile]: I think all release builds have the sse4.2 optimisations. The unportable bit is specific to RocksDB (maybe SSE4.2 there? I'm not sure)

Comment (brson) [reply Makefile]: Yes you are right. The release target turns on "jemalloc portable sse no-fail". unportable_turns of the "portable" feature, which I don't know the purpose of. I'll have to look into this and update the comments.

Comment (brson) [reply Makefile]: Interesting that there is no target that excludes the "sse" feature.

**Pull request merged**